substrate/primitives/state-machine/src/backend.rs

```diff
@@ -248,9 +245,9 @@ pub trait Backend<H: Hasher>: sp_std::fmt::Debug {
 	/// Does not include child storage updates.
 	fn storage_root<'a>(
 		&self,
-		delta: impl Iterator<Item = (&'a [u8], Option<&'a [u8]>)>,
+		delta: impl Iterator<Item = (&'a [u8], Option<&'a [u8]>, Option<ChildChangeset<H::Out>>)>,
 		state_version: StateVersion,
-	) -> (H::Out, BackendTransaction<H>)
+	) -> TrieCommit<H::Out>
```

Comment:
This is now tightly coupling every state machine used in Polkadot-SDK to the parity trie format.

I hope there is a way to avoid it.
I had planned to experiment with some next-gen trie-DBs that would be incompatible with this change.

Reply:
Yes, different backend requires an abstraction on the payload return by the state machine.
That is something I tried to achieve in the past (abstract state backend, abstract payload and abstract db storage), and never put it to completion (big refactoring, quite impacting at the time, maybe more doable today).
The issue with the existing code is that it abstracted only on H (the hasher), so you will need to implement it anyway.
This change  (TrieCommit) just change the existing static backend to the new one needed for new trie design: switching from PrefixedMemoryDb<H> + H::Out to TrieCommit<H::Out>. (old BackendTransaction looks like an abstraction but is just a type alias).
If the next-gen trie-db it is what is describe in your recent blog post, PrefixedMemoryDb<H> + H::Out is likely not what you need, neither is TrieCommit .
For experiment I would probably just use enum variants (db , statemachine triebackend and db payload (Backendtransaction). Making substrate generic over state backend would really be wonderful, but quite a refactoring.

Reply:
I could not find `trie_db::Changeset` anywhere, but anyways..

We'd want abstraction over data structures, not abstraction over tries.  We know optimal data structures should store data only at leaves, never at internal nodes, making tries a dead end.  We could envision several data structures worth exploring, but they're all binary trees with no data at internal nodes, aka no more tries.

Reply:
yeah that's not merge https://github.com/paritytech/trie/pull/210

Reply:
> Making substrate generic over state backend would really be wonderful, but quite a refactoring.

I am kind of surprised to hear this. At least, from the runtime perspective the interfaces look very general, which means that we just have to find the right boundary.
  
Maybe the `sp_state_machine` is the wrong place for that generalization, but it should be somewhere

Reply:
I guess we can use a new, substrate-defined type here.

The goal of this PR is to bring some optimizations to the existing trie format. Basically, for the DB-backed tries we want to be able to communicate the tree structure to the database level, as opposed to just pushing a set of nodes. 

For the next trie generation, I envision the connection between the trie and the DB will be even more integrated.

Reply:
Tries were always an incorrect design choice though. 

Intuitively this is clear:  If you have internal nodes then you expect to double PoV size over storing data only at leaves.

There is experemental evidence by others that the possible optimizations from internal nodes do not really pan out in pracitce.  You could do that sort of optimization with leaves anyways.   Also that radix 2 is always the correct choice when using symmetric hashing.

Now if you want a quick hack that redices PoV size for tries by replacing blake2 by blake3 compression plus binary merkle proofs, then one could do this.  I'd rather see something like what Rob proposes, meaning avoid tries entirely. 

Anyways we should not bake tries in any deeper.  If anything, our goals should be to depricate tries entirely.

Reply:
> Intuitively this is clear: If you have internal nodes then you expect to double PoV size over storing data only at leaves.

Not if the internal node should always be part of the pov, but yes not having internal node sounds like a good idea (even if I would prefer having single length key in this case (losing ordering is not really good: if ordering don't matter when key length differ, I am not sure there is a real use of having shorter keys).

I think everybody is thinking in the same direction here. (this PR is not related to it (it does not change anything about the trie, naming some branches new-trie is confusing I admit, it is really only a change for addressing merkle structure, which I think is not strictly incomaptible with a single state deterministic node address storage)).

Reply:
Anyone looked at https://github.com/cosmos/iavl yet?

Reply:
> For the next trie generation, I envision the connection between the trie and the DB will be even more integrated.

Yes, this seems very reasonable for maximum performance. Substrate should support some generality over storage backend. There are a few good options we have seen:
  * Jellyfish hex (small witness size, compact disk representation)
  * IAVL+ (balanced, hard to DoS, lower PoV usage)
  * Verkle (compact proofs, high compute costs, ZK friendly)
  * Binary Merkle (can be SSD optimized)

I personally suggest we should not assume that there is one "next generation" of the trie. Different options will work for different use-cases and architectures.